File: 50. Pow(x, n)/py/myPow.py
# LeetCode problem: 50. Pow(x, n)
# <https://leetcode.com/problems/powx-n/description/>

class Solution:
    def myPow(self, x: float, n: int) -> float:
        """
        Algorithm: Binary Exponentiation
        Time Complexity: O(log n)
        """
        if n == 0:
            return 1
        if n < 0:
            x, n = 1 / x, -n
        result = 1
        while n:
            if n % 2 == 1:
                result *= x
            x *= x
            n //= 2
        return result

File: 50. Pow(x, n)/cs/MyPow/MyPow.cs
// LeetCode problem: 50. Pow(x, n)
// https://leetcode.com/problems/powx-n/description/

namespace LeetCode.MyPow;

public class Solution
{
    public static double MyPow(double x, int n)
    {
        if (n == 0)
            return 1;
        long absN = Math.Abs((long)n);
        if (n < 0)
        {
            x = 1.0 / x;
            n = -n;
        }
        double result = 1.0;
        while (absN != 0)
        {
            if (absN % 2 == 1)
                result *= x;
            x *= x;
            absN /= 2;
        }
        return result;
    }
}